paka/kube_resources/model_group/models/abstract.py

```diff
@@ -0,0 +1,338 @@
+        """
+        with concurrent.futures.ThreadPoolExecutor(
+            max_workers=self.download_max_concurrency
+        ) as executor:
+            executor.map(self.download, urls, sha256s)
+
+    def upload_to_s3(
+        self, response: requests.Response, s3_file_name: str
+    ) -> tuple[Any, str]:
+        """
+        Uploads a single file to S3.
+
+        Args:
+            response (requests.Response): The response object from the file download.
+            s3_file_name (str): The name of the file in S3.
+
+        Returns:
+            tuple: A tuple containing the upload ID and the SHA256 hash of the file.
+        """
+        logger.info(f"Uploading model to {s3_file_name}")
+        sha256 = hashlib.sha256()
+        total_size = int(response.headers.get("content-length", 0))
+        processed_size = 0
+
+        upload = self.s3.create_multipart_upload(
+            Bucket=self.s3_bucket, Key=s3_file_name
+        )
+        upload_id = upload["UploadId"]
+        parts = []
+
+        with concurrent.futures.ThreadPoolExecutor(
+            max_workers=self.s3_max_concurrency
+        ) as executor:
+            futures: List[concurrent.futures.Future] = []
+            part_number = 1
+
+            for chunk in response.iter_content(chunk_size=self.s3_chunk_size):
+                sha256.update(chunk)
+                while len(futures) >= self.s3_max_concurrency:
+                    # Wait for one of the uploads to complete
+                    done, _ = concurrent.futures.wait(
+                        futures, return_when=concurrent.futures.FIRST_COMPLETED
+                    )
+                    for future in done:
+                        parts.append(future.result())
+                        futures.remove(future)
+
+                # Submit new chunk for upload
+                future = executor.submit(
+                    self.upload_part,
+                    s3_file_name,
+                    upload_id,
+                    part_number,
+                    chunk,
+                )
+                futures.append(future)
+                part_number += 1
+                processed_size += len(chunk)
+                progress = (processed_size / total_size) * 100
+                print(f"Progress: {progress:.2f}%", end="\r")
+
+            # Wait for all remaining uploads to complete
+            for future in concurrent.futures.as_completed(futures):
+                parts.append(future.result())
+
+        parts.sort(key=lambda part: part["PartNumber"])
+        self.s3.complete_multipart_upload(
+            Bucket=self.s3_bucket,
+            Key=s3_file_name,
+            UploadId=upload_id,
+            MultipartUpload={"Parts": parts},
+        )
+
+        logger.info(f"File uploaded to S3: {s3_file_name}")
+        sha256_value = sha256.hexdigest()
+        logger.info(f"SHA256 hash of the file: {sha256_value}")
+        return upload_id, sha256_value
+
+    def upload_fs_to_s3(
```

Comment:
This has a lot of duplicates with upload_to_s3. need consolidation

Reply:
we can remove upload_to_s3 method which is using requests http and not used anywhere.